hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -1271,84 +1307,54 @@ bool MDMParser::powerOff(void)
+
+    if (_cancel_all_operations) {
+        continue_cancel = true;
+        resume(); // make sure we can use the AT parser
+    }
+
+    check_ri = true;
+    if (!softPowerOff()) {
+        if (_dev.dev == DEV_SARA_R410) {
+            // If memory issue is present, ensure we don't force a power off too soon
+            // to avoid hitting the 124 day memory housekeeping issue, AT+CPWROFF will
+            // handle this delay automatically, or timeout after 40s.
+            if (_memoryIssuePresent) {
+                MDM_INFO("%s Modem not responsive, waiting up to 30s to power off with PWR_UC...", POWER_OFF_MSG);
+                system_tick_t now = HAL_Timer_Get_Milli_Seconds();
+                if (_timePowerOn == 0) {
+                    // fallback to max timeout of 30s to be safe
+                    _timePowerOn = now;
                 }
-                HAL_GPIO_Write(PWR_UC, 0);
-                // >1.5 seconds on SARA R410M
-                // >1 second on SARA U2
-                // plus a little extra for good luck
-                HAL_Delay_Milliseconds(1600);
-                HAL_GPIO_Write(PWR_UC, 1);
-                break;
-            } else {
-                sendFormated("AT+CPWROFF\r\n");
-                int ret = waitFinalResp(nullptr, nullptr, CPWROFF_TIMEOUT);
-                if (RESP_OK == ret) {
-                    if (_dev.dev == DEV_SARA_R410) {
-                        check_ri = true;
+                // check for timeout (VINT == LOW, Powered on 30s ago, Registered 20s ago)
+                do {
+                    now = HAL_Timer_Get_Milli_Seconds();
+                    // prefer to timeout 20s after registration if we are registered
+                    if (_timeRegistered) {
+                        if (now - _timeRegistered >= 20000UL) {
+                            break;
+                        }
+                    } else if (_timePowerOn && now - _timePowerOn >= 30000UL) {
+                        break;
                     }
-                    break;
-                } else if (RESP_ABORTED == ret) {
-                    MDM_INFO("%s found ABORTED, retrying...", POWER_OFF_MSG);
-                } else {
-                    MDM_INFO("%s timeout, retrying...", POWER_OFF_MSG);
-                }
+                    HAL_Delay_Milliseconds(100); // just wait
+                } while ( powerState() );
+                // reset timers
+                _timeRegistered = 0;
+                _timePowerOn = 0;
             }
         }
-    } else {
-        if (powerState()) {
-            check_ri = true;
-            MDM_INFO("%s modem is on unexpectedly. Turning it off...", POWER_OFF_MSG);
-            // Delay this long period so that the modem is fully on to accept the power-off sequence.
-            HAL_Delay_Milliseconds(12000);
-
+        // Skip power off sequence if power is already off
+        if (powerState() && _dev.dev != DEV_SARA_G350 && _dev.dev != DEV_UNKNOWN) {
```

Comment:
I think we should still attempt this even if `_dev.dev == DEV_UNKNOWN`. Again according to the datasheet, we shouldn't be able to accidentally power-on the modem, as power-on condition has a limit much lower than 1600ms.

Reply:
But in the datasheet it doesn't give the maximum low time on PWR_ON to power the modem. As long as the low time exceeds 5ms, the modem can be turned on.

Reply:
It does though: 
![image](https://user-images.githubusercontent.com/388478/97613577-6c7e0680-1a4b-11eb-8255-307d33a98f95.png)

Reply:
If I am reading this correctly, pulses over 80uS fall out of the timing requirements for power-on condition.

Reply:
I was talking about the SARA-G350:
<img width="878" alt="Screen Shot 2020-10-30 at 2 06 11 AM" src="https://user-images.githubusercontent.com/8130559/97614149-8bcd6180-1a54-11eb-991c-7783bb32b3af.png">

Reply:
Okay, double checked R410 as well and G350 is the only one without a maximum. We do have a `powerState()` check though, so supposedly we shouldn't be triggering a power-on, right?

Reply:
You're right 👍 . If `powerState()` is high for SARA-G350, it will leaks current anyway, even if we are powering on it again.

Reply:
And for the other unknown modules, except for the SARA-G350, they earn a chance to power off themselves.